src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,175 @@
+from pants.backend.scala.lint.scalafmt.skip_field import SkipScalafmtField
+from pants.backend.scala.lint.scalafmt.subsystem import ScalafmtSubsystem
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.testutil.rule_runner import logging
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ScalafmtFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafmtField).value
+
+
+class ScalafmtRequest(ScalaFmtRequest, LintRequest):
+    field_set_type = ScalafmtFieldSet
+
+
+class ScalafmtToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = ScalafmtSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: ScalafmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+@logging
+@rule(level=LogLevel.DEBUG)
+async def setup_google_java_format(
+    setup_request: SetupRequest,
+    tool: ScalafmtSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> Setup:
+    source_files, tool_classpath, conf_digest = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+        ),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(tool.resolved_lockfile(),),
+            ),
+        ),
+        # TODO: Figure out how to scan for the correct configuration file to use while also dealing with
+        #  the seeming requirement that `version` must be set in the configuration file.
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        path=".scalafmt.conf",
+                        content=textwrap.dedent(
+                            f"""\
+                version = "{tool.version}"
+                runner.dialect = scala213
```

Comment:
@stuhood / @chrisjrn: These configuration entries are required for `scalafmt` to operate. 

1. Should the scalafmt version come from the `.scalafmt.conf` file instead of the Pants option?
2. `scalafmt` looks by default either for a `.scalafmt.conf` in the same directory or in the root directory. Should we instead find the nearest ancestor `.scalafmt.conf`?

Reply:
My $0.02:

For part 1:

1. Have a pants setting for the version, which defaults to "use scafmt.conf please"
2. If the pants setting is not set, use scalafmt.conf

For part 2:

1. Have a pants setting for which scalafmt.conf file to use, with options "same directory", "root directory", "ancestor", or "specific file"
2. If not set, default to same directory only (until this appears to be an incorrect option).

Reply:
> Have a pants setting for the version, which defaults to "use scafmt.conf please"

The issue that I found there is that the version is an input to generating the tool lockfile. It could confuse users for changes to the version in scalafmt.conf to not take effect without regenerating the tool lockfile. Maybe we should just also require scalafmt.conf _somewhere_ in the hierarchy and let `scalafmt` make its own warnings about version mismatches?

Reply:
> Have a pants setting for which scalafmt.conf file to use, with options "same directory", "root directory", "ancestor", or "specific file"

Does it need to be configurable? The "find nearest ancestor" logic would subsume much of needing a config option since the user can just place the scalafmt.conf files as necessary for the desired effect.

The scalafmt default search logic (same directory or root of repo) doesn't seem very friendly to multi-team monoreos. The "find nearest ancestor" logic would actually be an improvement for monorepos.

Reply:
I think the "specific file" thing will prove important at some point, but the other two may not be 🤷

Reply:
In my experience, repos that use `scalafmt` all actually do have a `scalafmt.conf`, because there is so much configuration, and not very much convention.

> Maybe we should just also require scalafmt.conf somewhere in the hierarchy and let scalafmt make its own warnings about version mismatches?

I'm fine with this for now. Since the tool won't operate without the `.scalafmt.conf` present, you can set `GlobMatchErrorBehavior.error` while slurping it in, and then regex match to assert that the version matches our configured version or something.

Reply:
(also, there is prior art in that formatters on the Python side will slurp in `pyproject.toml` this way: see e.g. https://github.com/pantsbuild/pants/blob/3442d085c6906e143107241f3128a11078f9caa8/src/python/pants/backend/python/subsystems/pytest.py#L163-L178)

Reply:
There's also prior art with using a config file where we need to ensure a value is set a certain way:

https://github.com/pantsbuild/pants/blob/4c871edc37db2925f1d8647d5f16f272a923a4d3/src/python/pants/backend/python/goals/coverage_py.py#L327-L342